include/counter.h

```diff
@@ -27,30 +27,71 @@
 extern "C" {
 #endif
 
-typedef void (*counter_callback_t)(struct device *dev, void *user_data);
+/** @brief Single shot callback
+ *
+ * @param dev       Pointer to the device structure for the driver instance.
+ * @param id        Single shot request id.
+ * @param counter   Counter value that triggered the callback.
+ * @param user_data User data.
+ */
+typedef void (*counter_single_shot_callback_t)(struct device *dev,
+					       u32_t id,
+					       u32_t counter,
+					       void *user_data);
+
+struct counter_single_shot {
+	/** Index of the request. Maximal index is implementation dependent.*/
+	u8_t id;
+	/** If true ticks are treated as absolute value, else it is relative to
+	 * the counter reading performed during the call.
+	 */
+	bool abs;
+	u32_t ticks;
+	counter_single_shot_callback_t callback;
+	void *user_data;	
+};
+
+/* backward compatibility */
+#define counter_set_alarm counter_set_overflow
+
+typedef void (*counter_callback_t)(struct device *dev,
+					    void *user_data);
 
 typedef int (*counter_api_start)(struct device *dev);
 typedef int (*counter_api_stop)(struct device *dev);
 typedef u32_t (*counter_api_read)(struct device *dev);
-typedef int (*counter_api_set_alarm)(struct device *dev,
-				     counter_callback_t callback,
-				     u32_t count, void *user_data);
+typedef int (*counter_api_set_single_shot)(struct device *dev,
+					   struct counter_single_shot *params);
+
+typedef int (*counter_api_set_overflow)(struct device *dev,
+					u32_t ticks,
+					counter_callback_t callback,
+					void *user_data);
+
 typedef u32_t (*counter_api_get_pending_int)(struct device *dev);
+typedef int (*counter_api_ms_to_ticks)(struct device *dev,
+				       u32_t ms,
+				       u32_t *ticks );
+typedef int (*counter_api_us_to_ticks)(struct device *dev,
```

Comment:
I would just return the ticks. I don't see any error that could happen (maybe I am wrong)

Reply:
returning ticks would simplify conversion. The only thing that could go wrong is when requested ms/us does not fit in 32b ticks. In that case function would saturate at max ticks. Of course, having ticks as return value has 'style' advantage ( counter_set_overflow(... counter_us_to_ticks(1000))).

Reply:
@tbursztyka: We will need reverse functions for precise time measurement. We can use k_cycle_32_get() for that, but on some architectures this is bound to the most energy efficient timer (read: low frequency). There are faster timers in the system, but there is no official API for controlling them.